### PR TITLE
Allow CeleryExecutor to "adopt" an orphaned queued or running task

### PR DESCRIPTION
**Pull request by ashb**
This can happen when a task is enqueued by one executor, and then that
scheduler dies/exits.

The default fallback behaviour is unchanged -- that queued tasks are
cleared and then and then later rescheduled.

But for Celery we can do better -- if we record the Celery-generated
task_id, we can then re-create the AsyncResult objects for orphaned
tasks at a later date.

However since Celery just reports all AsyncResult as "PENDING", even if
they aren't tasks currently in the broker queue, we need to apply a
timeout to "unblock" these tasks in case they never actually made it to
the Celery broker.

This all means that we can adopt tasks that have been enqueued another
CeleryExecutor if it dies, without having to clear the task and slow
down. This is especially useful as the task may have already started
running, and while clearing it would stop it, it's better if we don't
have to reset it!

Part of #9630

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/executors/celery_executor.py
Comment (XD-DENG) [inline airflow/executors/celery_executor.py:373]: Maybe we do not need to have `tis_by_celery_task_id = {ti.external_executor_id: ti for ti in tis}` (line 360) separately. Instead, we can have 
```python
celery_tasks = {
    ti.external_executor_id: (AsyncResult(ti.external_executor_id), ti)
    for ti in tis
    if ti.external_executor_id is not None
}
```

Then lines 369-370 can also be merged into
```python
result, ti = celery_tasks[celery_task_id]
```

By doing this, we can avoid traversing `tis` for two times (only one time would suffice), also save one query on the dict.

**Review by XD-DENG — COMMENTED** [review]
A minor suggestion

File: airflow/executors/celery_executor.py
Comment (XD-DENG) [reply airflow/executors/celery_executor.py:373]: In addition, `tis_by_celery_task_id` covers all `tis`, which is not necessary, because it will only be used to query `tis` which present in `celery_tasks` (essentially means `tis` whose `external_executor_id` is not `None`).

So my suggested revision above may save a little memory as well ;-)

File: airflow/executors/celery_executor.py
Comment (XD-DENG) [inline airflow/executors/celery_executor.py]: `see which if the tis...` may contain a typo?

File: airflow/migrations/versions/e1a11ece99cc_add_external_executor_id_to_ti.py
Comment (XD-DENG) [inline airflow/migrations/versions/e1a11ece99cc_add_external_executor_id_to_ti.py:40]: `external_executor_id` is the [Celery task's UUID](https://docs.celeryproject.org/en/stable/reference/celery.result.html#celery.result.AsyncResult.id) if I'm not wrong. Is 250 too big for it?

File: airflow/executors/celery_executor.py
Comment (ashb) [reply airflow/executors/celery_executor.py]: Yup, `which of` it should be

File: airflow/migrations/versions/e1a11ece99cc_add_external_executor_id_to_ti.py
Comment (ashb) [reply airflow/migrations/versions/e1a11ece99cc_add_external_executor_id_to_ti.py:40]: To big for Celery, but this gives a bit more freedom for other executors to store things in there.

Comment (XD-DENG) [reply airflow/migrations/versions/e1a11ece99cc_add_external_executor_id_to_ti.py:40]: Get your point. 

IMHO possibly somewhere around 100 would suffice given its nature as `id`. But feel free to ignore this one.

File: airflow/executors/celery_executor.py
Comment (ashb) [reply airflow/executors/celery_executor.py:373]: Nice!

Comment (ashb) [reply airflow/executors/celery_executor.py:373]: Done.

**Comment by ashb** [comment]
The doc tests are now failing with:

> /opt/airflow/docs/_api/airflow/executors/base_executor/index.rst:198:more than one target found for cross-reference 'TaskInstance': airflow.models.TaskInstance, airflow.models.taskinstance.TaskInstance

But I didn't change the imports in that file :/

File: airflow/migrations/versions/e1a11ece99cc_add_external_executor_id_to_ti.py
Comment (XD-DENG) [reply airflow/migrations/versions/e1a11ece99cc_add_external_executor_id_to_ti.py:40]: Just realise `ID_LEN` is specified in `airflow.models.base` as a constant. Yes, please ignore this comment

File: airflow/executors/base_executor.py
Comment (XD-DENG) [inline airflow/executors/base_executor.py]: Minor: another typo spotted, `addopt`.

**Review by XD-DENG — CHANGES_REQUESTED** [review]

File: airflow/config_templates/config.yml
Comment (turbaszek) [inline airflow/config_templates/config.yml]:
```suggestion
        Time in seconds after which Adopted tasks are cleared by CeleryExecutor. This is helpful to clear
```
Just to clarify

File: airflow/executors/celery_executor.py
Comment (turbaszek) [inline airflow/executors/celery_executor.py]: Not a big thing but we can do one `for` loop instead of this dict comprehension and list in L365, right?

File: airflow/executors/celery_executor.py
Comment (turbaszek) [inline airflow/executors/celery_executor.py]:
```suggestion
            result.backend = cached_celery_backend
```
Duplicated line

File: airflow/executors/celery_executor.py
Comment (turbaszek) [inline airflow/executors/celery_executor.py]:
```suggestion
            adopted.append(f"{ti}  in state {state}")
```
How about this?

File: airflow/executors/celery_executor.py
Comment (ashb) [reply airflow/executors/celery_executor.py]: We could, but they have different conditions (`if` and `if not`) - we'd need to do something like:

```python
celery_tasks = []
not_adopted = []

for ti in its:
  if ti.external_executor_id:
    celery_tasks.append(...)
  else:
    not_adopted.append(...)
```

Is that what you meant?

File: airflow/jobs/scheduler_job.py
Comment (turbaszek) [inline airflow/jobs/scheduler_job.py:1562]: Not related, but is this comment still valid?

File: airflow/utils/sqlalchemy.py
Comment (turbaszek) [inline airflow/utils/sqlalchemy.py]:
```suggestion
def skip_locked(session: Session, of: Any = None) -> Dict[str, Any]:
```
Can you describe the `of` or use something more informative? I'm not sure if I got it 😄

File: airflow/executors/celery_executor.py
Comment (turbaszek) [reply airflow/executors/celery_executor.py]: Yes, I don't think this changes performance

File: airflow/utils/sqlalchemy.py
Comment (ashb) [reply airflow/utils/sqlalchemy.py]: I wasn't aware of any more concrete type -- it's "any SQLA Model class". I'll see what I can find.

Comment (turbaszek) [reply airflow/utils/sqlalchemy.py]: My concern is not about typing but about the name of argument, the `of` is not meaningful to me

Comment (ashb) [reply airflow/utils/sqlalchemy.py]: Oh, right. That is mirroring the SQLA argument https://docs.sqlalchemy.org/en/13/orm/query.html?highlight=with_for_update#sqlalchemy.orm.query.Query.with_for_update:

> sqlalchemy.orm.query.Query.with_for_update(read=False, nowait=False, of=None, skip_locked=False, key_share=False)

Comment (ashb) [reply airflow/utils/sqlalchemy.py]: Oh wait, I see a way to remove this option from here:

`.with_for_update(**skip_locked(of=DagRun, session=session)` -> `with_for_update(of=DagRun, **skip_locked(session))`

File: airflow/executors/celery_executor.py
Comment (ashb) [reply airflow/executors/celery_executor.py]: :+1:

File: airflow/utils/sqlalchemy.py
Comment (ashb) [reply airflow/utils/sqlalchemy.py]: Fixed in https://github.com/apache/airflow/pull/10949/commits/3f4b2ea0c45541f462ad102646d685a74416aba2

File: airflow/executors/celery_executor.py
Comment (ashb) [reply airflow/executors/celery_executor.py]: Done in https://github.com/apache/airflow/pull/10949/commits/e4f915efb2b1bd4165d33425799f3bbffd88caa3

File: airflow/executors/celery_executor.py
Comment (ashb) [reply airflow/executors/celery_executor.py]: https://github.com/apache/airflow/pull/10949/commits/e4f915efb2b1bd4165d33425799f3bbffd88caa3

File: airflow/executors/celery_executor.py
Comment (ashb) [reply airflow/executors/celery_executor.py]: https://github.com/apache/airflow/pull/10949/commits/e4f915efb2b1bd4165d33425799f3bbffd88caa3

File: airflow/config_templates/config.yml
Comment (ashb) [reply airflow/config_templates/config.yml]: https://github.com/apache/airflow/pull/10949/commits/e4f915efb2b1bd4165d33425799f3bbffd88caa3

File: airflow/jobs/scheduler_job.py
Comment (ashb) [reply airflow/jobs/scheduler_job.py:1562]: Not sure off the top of my head.

**Review by turbaszek — APPROVED** [review]
Looks good to me 👍

File: airflow/executors/celery_executor.py
Comment (XD-DENG) [inline airflow/executors/celery_executor.py]: Only one more minor comment relating to naming: `self. adopted_task_timeouts` and `adopted_task_timeout` may make this less comprehensive.

I understand they are linked via `self.adopted_task_timeouts[ti.key] = ti.queued_dttm + self.adoption_timeout`, but the extremely similar names made me think they are "directly" linked to each other.

I would suggest to rename either one.

Otherwise all look good to me.

**Review by XD-DENG — COMMENTED** [review]
One more minor comment.

File: airflow/executors/celery_executor.py
Comment (ashb) [reply airflow/executors/celery_executor.py]: Good shout. I'll rename `adopted_task_timeouts` to `adopted_task_by_timeout`? `abandon_adopted_tasks_after`?

**Review by XD-DENG — APPROVED** [review]

File: airflow/executors/celery_executor.py
Comment (ashb) [reply airflow/executors/celery_executor.py]: Or perhaps `adopted_task_timeout` to `task_adoption_timeout` (and leave `adopted_task_timeouts`)

Comment (XD-DENG) [reply airflow/executors/celery_executor.py]: I leave this to an English speaker 😉  I will be happy as long as they distinguish well from each other. 

But my preference goes for `adopted_task_timeout`-> `task_adoption_timeout`

**Comment by ashb** [comment]
Right, final rebase done, will merge once tests are green.

(I wish Github had a button for that)

**Pull request merged**
